controllers/oscmachine_keypair_controller.go

```diff
@@ -0,0 +1,111 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	infrastructurev1beta1 "github.com/outscale-dev/cluster-api-provider-outscale.git/api/v1beta1"
+	"github.com/outscale-dev/cluster-api-provider-outscale.git/cloud/scope"
+	"github.com/outscale-dev/cluster-api-provider-outscale.git/cloud/services/security"
+	tag "github.com/outscale-dev/cluster-api-provider-outscale.git/cloud/tag"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func checkKeypairFormatParameters(machineScope *scope.MachineScope) (string, error) {
+	machineScope.Info("Check Keypair parameters")
+
+	var keypairSpec *infrastructurev1beta1.OscKeypair
+	nodeSpec := machineScope.GetNode()
+	if nodeSpec.KeyPair.Name == "" {
+		nodeSpec.SetKeyPairDefaultValue()
+		keypairSpec = &nodeSpec.KeyPair
+	} else {
+		keypairSpec = machineScope.GetKeypair()
+	}
+
+	keypairName := keypairSpec.Name + "-" + machineScope.GetUID()
+	keypairTagName, err := tag.ValidateTagNameValue(keypairName)
+	if err != nil {
+		return keypairTagName, err
+	}
+
+	publicIp := keypairSpec.PublicKey
+	_, err = security.ValidateKeypairPublicIp(publicIp)
+	if err != nil {
+		return keypairTagName, err
+	}
+	return "", nil
+}
+
+// getKeyPairResourceId return the keypairName from the resourceMap base on resourceName (tag name + cluster uid)
+func getKeyPairResourceId(resourceName string, machineScope *scope.MachineScope) (string, error) {
+	keypairRef := machineScope.GetKeypairRef()
+	if keypairName, ok := keypairRef.ResourceMap[resourceName]; ok {
+		return keypairName, nil
+	} else {
+		return "", fmt.Errorf("%s does not exist", resourceName)
+	}
+}
+
+// reconcileKeypair reconcile the keypair of the machine
+func reconcileKeypair(ctx context.Context, machineScope *scope.MachineScope, keypairSvc security.OscKeyPairInterface) (reconcile.Result, error) {
+	machineScope.Info("Create Keypair or add existing one")
+	var keypairSpec *infrastructurev1beta1.OscKeypair
+	keypairSpec = machineScope.GetKeypair()
+	keypairRef := machineScope.GetKeypairRef()
+	keypairName := keypairSpec.Name + "-" + machineScope.GetUID()
+
+	if len(keypairRef.ResourceMap) == 0 {
+		keypairRef.ResourceMap = make(map[string]string)
+	}
+	if keypairSpec.ResourceId != "" {
+		keypairRef.ResourceMap[keypairName] = keypairSpec.ResourceId
+	}
+
+	if _, err := keypairSvc.GetKeyPair(keypairName); err != nil {
```

Comment:
here if getKeyPair return null you will  create keypair.
here if getKeyPair return a keypair you will  create keypair.
(Please change it like this ?)

Reply:
```
 keypair, err := keypairSvc.GetKeyPair(keypairName)
 if err != nil {
      return reconcile.Result{}, err
}
if keypair == nil || keypairSpec.ResourceId == "" {
goto CREATE
}

if keypairSpec.PublicKey != "" {
 security.ValidateKeypairPublicIp(keypairSpec.PublicKey) 
 _, err := security.ValidateKeypairPublicIp(keypairSpec.PublicKey)
 if err != nil {
      return reconcile.Result{}, err
}
 
```

Reply:
done